src/Microsoft.TemplateEngine.Utils/Glob.cs

```diff
@@ -0,0 +1,275 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.TemplateEngine.Utils
+{
+    public class Glob
+    {
+        private readonly IReadOnlyList<IMatcher> _matchers;
+
+        private Glob(IReadOnlyList<IMatcher> matchers)
+        {
+            _matchers = matchers;
+        }
+
+        private interface IMatcher
+        {
+            int MinConsume { get; }
+
+            bool ProducesCheckpoint { get; }
+
+            bool CanConsume(string test, int startAt, out int endPosition);
+        }
+
+        public static Glob Parse(string pattern)
+        {
+            List<IMatcher> matchers = new List<IMatcher>();
+
+            for (int i = 0; i < pattern.Length; ++i)
+            {
+                switch (pattern[i])
+                {
+                    case '[':
+                        List<char> values = new List<char>();
+                        for (; i < pattern.Length; ++i)
+                        {
+                            if (pattern[i] == ']')
+                            {
+                                break;
+                            }
+
+                            values.Add(pattern[i]);
+                        }
+                        matchers.Add(new LiteralMatcher(values));
+                        break;
+                    case '*':
+                        if (pattern.Length > i + 1 && pattern[i + 1] == '*')
+                        {
+                            if (i < pattern.Length - 2 && pattern[i + 2] == '/')
+                            {
+                                matchers.Add(new ExactPathMatcher());
+                                i += 2;
+                            }
+                            else
+                            {
+                                matchers.Add(new PathMatcher());
+                                ++i;
+                            }
+                        }
+                        else
+                        {
+                            matchers.Add(new NameMatcher());
+                        }
+                        break;
+                    default:
+                        matchers.Add(new LiteralMatcher(pattern[i]));
+                        break;
+                }
+            }
+
+            return new Glob(matchers);
+        }
+
+        public bool IsMatch(string test)
+        {
+            Stack<Checkpoint> checkpoints = new Stack<Checkpoint>();
+
+            int currentMatcher = 0;
+            int i = 0;
+
+            //See if we can just do a name match
+            if (!_matchers.Any(x => x is PathMatcher || x is ExactPathMatcher || (x as LiteralMatcher)?.Char?.FirstOrDefault() == '/'))
```

Comment:
This could be evaluated once in the constructor and the result stored in a private bool, to be checked here, instead of rechecking each time.